### PR TITLE
Laboratory work #4, Semenova Taisiya - 23FPL2

### PR DESCRIPTION
**Pull request by semtassta**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Looks strange. Did you pass the 3rd lab?

**Comment by semtassta** [comment]
> Looks strange. Did you pass the 3rd lab?

Not yet

File: lab_4_retrieval_w_clustering/main.py [hunk 1]
```diff
@@ -343,6 +517,17 @@ def _is_convergence_reached(
         Returns:
             bool: True if the distance is correct, False in other cases.
         """
+        if not self.__clusters:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 2]
```diff
@@ -343,6 +517,17 @@ def _is_convergence_reached(
...
+        for index, old_cluster in enumerate(self.__clusters):
+            new_centroid = new_clusters[index].get_centroid()
+            old_centroid = old_cluster.get_centroid()
+            distance = calculate_distance(new_centroid, old_centroid)
+            if distance is None:
+                raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 3]
```diff
@@ -305,6 +449,36 @@ def infer(self, query_vector: Vector, n_neighbours: int) -> list[tuple[float, in
         Returns:
             list[tuple[float, int]]: Distance to relevant document and document index.
         """
+        if (not isinstance(query_vector, tuple) or not query_vector or
+                not isinstance(n_neighbours, int) or n_neighbours < 1):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 4]
```diff
@@ -305,6 +449,36 @@ def infer(self, query_vector: Vector, n_neighbours: int) -> list[tuple[float, in
...
+        centroids = [cluster.get_centroid() for cluster in self.__clusters]
+        closest_centroid = centroids[0]
+        distance_minimal = calculate_distance(query_vector, centroids[0])
+        if distance_minimal is None:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 5]
```diff
@@ -305,6 +449,36 @@ def infer(self, query_vector: Vector, n_neighbours: int) -> list[tuple[float, in
...
+        for centroid in centroids:
+            distance = calculate_distance(query_vector, centroid)
+            if distance is None:
+                raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 6]
```diff
@@ -305,6 +449,36 @@ def infer(self, query_vector: Vector, n_neighbours: int) -> list[tuple[float, in
...
+                distance_minimal = distance
+                closest_centroid = centroid
+        closest_index = centroids.index(closest_centroid)
+        closest_cluster = self.__clusters[closest_index]
+        indices = closest_cluster.get_indices()
+        vectors = self._db.get_vectors(indices)
+        documents_relevant = []
+        distance_vector_minimal = calculate_distance(query_vector, centroids[0])
+        if distance_vector_minimal is None:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 7]
```diff
@@ -305,6 +449,36 @@ def infer(self, query_vector: Vector, n_neighbours: int) -> list[tuple[float, in
...
+        for index, vector in vectors:
+            distance_vector = calculate_distance(query_vector, vector)
+            if distance_vector is None:
+                raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 8]
```diff
@@ -287,6 +407,30 @@ def run_single_train_iteration(self) -> list[ClusterDTO]:
...
+        centroids = []
+        for cluster in self.__clusters:
+            cluster.erase_indices()
+            centroids.append(cluster.get_centroid())
+        vectors = self._db.get_vectors()
+        for vector_index, vector in vectors:
+            closest_centroid = centroids[0]
+            distance_minimal = calculate_distance(vector, centroids[0])
+            if distance_minimal is None:
+                raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 9]
```diff
@@ -287,6 +407,30 @@ def run_single_train_iteration(self) -> list[ClusterDTO]:
...
+            for centroid in centroids:
+                distance = calculate_distance(vector, centroid)
+                if distance is None:
+                    raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 10]
```diff
@@ -182,6 +266,22 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
...
+            raise ValueError
+        if not isinstance(n_neighbours, int) or n_neighbours < 0:
+            raise ValueError
+        document_vectors = [vectors[1] for vectors in self._db.get_vectors()]
+        query_vectorized = self._index_document(query)
+        if not isinstance(query_vectorized, tuple):
+            raise ValueError
+        knn = self._calculate_knn(query_vectorized, document_vectors, n_neighbours)
+        if not isinstance(knn, list):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 11]
```diff
@@ -182,6 +266,22 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
...
             list[tuple[float, str]]: Relevant documents with their distances.
         """
+        if not query or not n_neighbours:
+            raise ValueError
+        if not isinstance(n_neighbours, int) or n_neighbours < 0:
+            raise ValueError
+        document_vectors = [vectors[1] for vectors in self._db.get_vectors()]
+        query_vectorized = self._index_document(query)
+        if not isinstance(query_vectorized, tuple):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 12]
```diff
@@ -182,6 +266,22 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
         Returns:
             list[tuple[float, str]]: Relevant documents with their distances.
         """
+        if not query or not n_neighbours:
+            raise ValueError
+        if not isinstance(n_neighbours, int) or n_neighbours < 0:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 13]
```diff
@@ -182,6 +266,22 @@ def retrieve_relevant_documents(self, query: str, n_neighbours: int) -> list[tup
         Returns:
             list[tuple[float, str]]: Relevant documents with their distances.
         """
+        if not query or not n_neighbours:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 14]
```diff
@@ -113,6 +162,31 @@ def put_corpus(self, corpus: Corpus) -> None:
...
+        if not isinstance(corpus, list):
+            raise ValueError
+        documents = []
+        vectors = {}
+        corpus_tokenized = []
+        for text in corpus:
+            text_tokenized = self._tokenizer.tokenize(text)
+            if text_tokenized is None:
+                raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 15]
```diff
@@ -113,6 +162,31 @@ def put_corpus(self, corpus: Corpus) -> None:
                 or if input arguments are empty,
                 or if methods used return None.
         """
+        if not corpus:
+            raise ValueError
+        if not isinstance(corpus, list):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 16]
```diff
@@ -113,6 +162,31 @@ def put_corpus(self, corpus: Corpus) -> None:
                 or if input arguments are empty,
                 or if methods used return None.
         """
+        if not corpus:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 17]
```diff
@@ -113,6 +162,31 @@ def put_corpus(self, corpus: Corpus) -> None:
...
+            if text_tokenized:
+                corpus_tokenized.append(text_tokenized)
+                documents.append(text)
+        self._vectorizer.set_tokenized_corpus(corpus_tokenized)
+        self._vectorizer.build()
+        i = 0
+        for text_tokenized in corpus_tokenized:
+            text_vectorized = self._vectorizer.vectorize(text_tokenized)
+            if text_vectorized is None:
+                raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 18]
```diff
@@ -67,6 +89,14 @@ def vectorize(self, tokenized_document: list[str]) -> Vector:
         Returns:
             Vector: BM25 vector for document.
         """
+        if not tokenized_document:
+            raise ValueError("Input argument is empty")
+        if not isinstance(tokenized_document, list):
+            raise ValueError("Inappropriate type input argument")
+        bm25 = self._calculate_bm25(tokenized_document)
+        if bm25 is None:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 19]
```diff
@@ -81,6 +111,21 @@ def _calculate_bm25(self, tokenized_document: list[str]) -> Vector:
...
+            raise ValueError("Input argument is empty")
+        if not isinstance(tokenized_document, list):
+            raise ValueError("Inappropriate type input argument")
+        if not self._vocabulary:
+            return ()
+        vector_bm25 = [0.0]*len(self._vocabulary)
+        dict_bm25 = calculate_bm25(self._vocabulary, tokenized_document, self._idf_values,
+                                   avg_doc_len=self._avg_doc_len, doc_len=len(tokenized_document))
+        if dict_bm25 is None:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_retrieval_w_clustering/main.py [hunk 20]
```diff
@@ -4,8 +4,16 @@
 Vector search with clusterization
 """
 
+from lab_2_retrieval_w_bm25.main import calculate_bm25
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: All imports together

File: lab_4_retrieval_w_clustering/main.py [hunk 21]
```diff
@@ -40,6 +53,9 @@ def __init__(self) -> None:
         """
         Initialize an instance of the BM25Vectorizer class.
         """
+        self._corpus = []
+        super().__init__(self._corpus)
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: super should be the first line in this method

File: lab_4_retrieval_w_clustering/main.py [hunk 22]
```diff
@@ -227,11 +331,15 @@ def set_new_centroid(self, new_centroid: Vector) -> None:
             ValueError: In case of inappropriate type input arguments,
                 or if input arguments are empty.
         """
+        if not new_centroid or new_centroid is None or not isinstance(new_centroid, tuple):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 23]
```diff
@@ -244,6 +352,10 @@ def add_document_index(self, index: int) -> None:
             ValueError: In case of inappropriate type input arguments,
                 or if input arguments are empty.
         """
+        if not isinstance(index, int) or not index or index < 0 or isinstance(index, bool):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Add message

File: lab_4_retrieval_w_clustering/main.py [hunk 24]
```diff
@@ -287,6 +407,30 @@ def run_single_train_iteration(self) -> list[ClusterDTO]:
...
+                if distance < distance_minimal:
+                    distance_minimal = distance
+                    closest_centroid = centroid
+            self.__clusters[centroids.index(closest_centroid)].add_document_index(vector_index)
+        for cluster in self.__clusters:
+            cluster_vectors = [vectors[index][1] for index in cluster.get_indices()]
+            centroid_updated = (sum(cluster_vectors[index]) / len(cluster_vectors)
+                                for index in range(len(cluster_vectors)-1))
+            cluster.set_new_centroid(tuple(centroid_updated))
+        return self.__clusters
```
Comment (artyomtugaryov) [inline lab_4_retrieval_w_clustering/main.py]: Are you sure that you need to create new  clustrs in `self.__clusters`? Probably you need create it in another variable and return

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_retrieval_w_clustering/main.py (hunk 1 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 2 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 3 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 4 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 5 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 15 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 16 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 17 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 18 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 19 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 22 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 23 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 21 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 20 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: done

File: lab_4_retrieval_w_clustering/main.py (hunk 14 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 13 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 12 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 11 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 10 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 9 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 8 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 7 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

File: lab_4_retrieval_w_clustering/main.py (hunk 6 above)
Comment (semtassta) [reply lab_4_retrieval_w_clustering/main.py]: Done

**Pull request closed without merging**